File: leetcode/group-anagrams.java
//https://leetcode.com/problems/group-anagrams/

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, Integer> map = new HashMap<String, Integer>();
        List<List<String>> groups = new ArrayList<List<String>>();

        for (String i : strs) {
            char[] str = i.toCharArray();
            Arrays.sort(str);
            String key = new String(str);
            Integer index = map.get(key);
            if (index != null) {
                groups.get(index).add(i);
            } else {
                map.put(key, groups.size());
                List<String> list = new ArrayList<String>();
                list.add(i);
                groups.add(list);
            }
        }

        return groups;
    }
}

File: leetcode/group_anagrams.py
# https://leetcode.com/problems/group-anagrams/


class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        grouped: dict[str] = {}

        for s in strs:
            key = "".join(sorted(s))
            if not grouped.get(key):
                grouped[key] = []
            grouped[key].append(s)

        out: list[list[str]] = []
        for i in grouped.keys():
            out.append(grouped.get(i))
        return out


class SolutionTwo:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        groups: dict[str, list[str]] = {}

        for string in strs:
            key: str = str(sorted(string))
            if not groups.get(key):
                groups[key] = []
            groups[key].append(string)

        result: list[list[str]] = []
        for val in groups.values():
            result.append(val)

        return result
